src/main/java/com/expertsoft/tasks/OrderStats.java

```diff
@@ -1,120 +1,148 @@
 import com.expertsoft.util.AveragingBigDecimalCollector;
 
 import java.math.BigDecimal;
+import java.util.Collection;
+import java.util.Collections;
 import java.util.List;
 import java.util.Map;
+import java.util.NoSuchElementException;
 import java.util.Optional;
+import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 /**
  * This class provides several methods to collect statistical information from customers and orders of an e-shop.
  * Your task is to implement methods of this class using Java 8 Stream API.
  * Each method is covered with a number of unit tests. You can use <code>mvn test</code> call to check your implementation.
- *
+ * <p>
  * Refer to the <code>com.expertsoft.model</code> package to observe the domain model of the shop.
  */
 class OrderStats {
 
     /**
      * Task 1 (⚫⚫⚪⚪⚪)
-     *
+     * <p>
      * Given a stream of customers, return the list of orders, paid using provided credit card type (Visa or MasterCard).
      *
      * @param customers stream of customers
-     * @param cardType credit card type
+     * @param cardType  credit card type
      * @return list, containing orders paid with provided card type
      */
     static List<Order> ordersForCardType(final Stream<Customer> customers, PaymentInfo.CardType cardType) {
-        return null;
+        return customers
+                .flatMap(customer -> customer.getOrders().stream().filter(order -> order.getPaymentInfo().getCardType().equals(cardType)))
+                .collect(Collectors.toList());
     }
 
     /**
      * Task 2 (⚫⚫⚪⚪⚪)
-     *
+     * <p>
      * Given a stream of orders, return a map, where keys are different order sizes and values are lists of orders,
      * referring to this sizes. Order size here is just a total number of products in the order.
      *
      * @param orders stream of orders
      * @return map, where order size values mapped to lists of orders
      */
     static Map<Integer, List<Order>> orderSizes(final Stream<Order> orders) {
-        return null;
+        Collection<Order> orderCollection = orders.collect(Collectors.toList());
+        return orderCollection.stream().collect(Collectors.groupingBy(order -> order.getOrderItems().size(), Collectors.toList()));
     }
 
     /**
      * Task 3 (⚫⚫⚫⚪⚪)
-     *
+     * <p>
      * Given a stream of orders, return true only if EVERY order in the stream contains at least
      * one product of the provided color and false otherwise.
      *
      * @param orders stream of orders
-     * @param color product color to test
+     * @param color  product color to test
      * @return boolean, representing if every order in the stream contains product of specified color
      */
     static Boolean hasColorProduct(final Stream<Order> orders, final Product.Color color) {
-        return null;
+        return orders.allMatch(order -> order.getOrderItems().stream().anyMatch(orderItem -> orderItem.getProduct().getColor().equals(color)));
```

Comment:
this method should check that all items of orders contain at least one product of the provided color
anyMatch method will return true even if just one order in  stream contains product of this color and this is not correct for current task

Reply:
no, according to req: "every order contains at least one product" so, I've performed allMatch for items and anyMatch for products. I think this is proper implementation )